lensless/mask.py

```diff
@@ -0,0 +1,321 @@
+
+        Parameters
+        ----------
+        sensor_size_px: tuple (dim=2)
+            size of the sensor (px)
+        sensor_size_m: tuple (dim=2)
+            size of the sensor (m)
+        feature_size: float
+            size of the feature (m)
+        distance_sensor: float
+            distance between the mask and the sensor (m)
+        wavelength: float
+            wavelength (m)
+        """
+        
+        self.mask = None
+        self.psf = None
+        self.phase_mask = None
+        self.height_map = None
+        self.sensor_size_px = sensor_size_px
+        self.sensor_size_m = sensor_size_m
+        self.feature_size = feature_size
+        self.distance_sensor = distance_sensor
+        self.wavelength = wavelength
+        self.create_mask()
+        self.compute_psf()
+    
+
+    @abc.abstractmethod
+    def create_mask(self):
+        """
+        Abstract mask creation method.
+        Creating mask with subclass-specific function.
+        """
+        pass
+
+
+    def compute_psf(self):
+        """
+        Computing the PSF.
+        Common to all types of masks.
+        """
+        self.psf, _, _ = angular_spectrum(
+            u_in=self.mask,
+            wv=self.wavelength,
+            d1=self.feature_size,
+            dz=self.distance_sensor,
+            dtype=np.float32
+        )
+        pass
+    
+
+    def shape(self):
+        """
+        Shape of the mask.
+        """
+        return self.mask.shape
+    
+
+    def phase_retrieval(self, lambd, d1, dz, n=1.5, n_iter=10):
+        """
+        Iterative phase retrieval algorithm from the PhlatCam article (https://ieeexplore.ieee.org/document/9076617)
+
+        Parameters
+        ----------
+        lambd: float
+            wavelength (m)
+        d1: float
+            sample period on the sensor i.e. pixel size (m)
+        dz: float
+            propagation distance between the mask and the sensor
+        n: float
+            refractive index of the mask substrate
+        n_iter: int
+            number of iterations
+        """
+        M_p = np.sqrt(self.psf)
+        for _ in progressbar.ProgressBar()(range(n_iter)):
+            M_phi = np.exp(1j * np.angle(fresnel_conv(M_p, lambd, d1, -dz, dtype=np.float32)[0]))
+            M_p = np.sqrt(self.psf) * np.exp(1j * fresnel_conv(M_phi, lambd, d1, dz, dtype=np.float32)[0])    
+        phi = np.angle(M_phi)
+
+        self.phase_mask = phi
+        self.height_map = lambd * phi / (2 * np.pi * (n-1))
+        pass
+
+
+
+
+
+class CodedAperture(Mask):
+    """
+    Coded aperture subclass of the Mask class
+    From the FlatCam article https://arxiv.org/abs/1509.00116
+    """
+
+    def __init__(self, 
+                 sensor_size_px: tuple, 
+                 sensor_size_m: tuple, 
+                 feature_size: tuple, 
+                 distance_sensor: float,
+                 wavelength: float, 
+                 method: str, 
+                 n_bits: int) -> None:
```

Comment:
could you add `**kwargs` to the constructor, as [here](https://github.com/LCAV/LenslessPiCam/blob/d7b743e56f975b472de006bd85d2298f0a71867c/lensless/gradient_descent.py#L53). And pass them to the parent constructor below. 

This way you can remove all mask-common parameters from this constructor, i.e. leaving only `method` and `n_bits` for this one.

Reply:
and how about defaulting `method` and `n_bits` to values from the paper?